### ISSUE TITLE
Sometimes the same fragment is added to a query twice

### ISSUE THREAD

**Issue by richburdon** [apollostack/apollo-client#897]
I have a parent container List that renders a list of child container Items. I also have pagination working (i.e., a fetchMore property that implements an updateQuery function).

However, if I change the value of one of the query variables (e.g., a "text" filter) then paging causes an error (the Item fragment is included twice as below).

From my experience with Relay, the parent query variables should be used to parameterize the fragment, but I don't know if that's relevant here, or possible (e.g., ${Item.getFragment('item', { filter: variables.filter })})

```
ApolloError: "GraphQL error: There can only be one fragment named "ItemFragment".

query GetItems($text: String, $offset: Int, $count: Int) {
  items(text: $text, offset: $offset, count: $count) {
    id
    ...ItemFragment
    __typename
  }
}

fragment ItemFragment on Item {
  id
  title
  __typename
}

fragment ItemFragment on Item { // Only happens when $text is changed.
  id
  title
  __typename
}
```

Unrelated, but one disadvantage of the convention to put fragments as static members of the component is you can't reference these from static propTypes {} member functions.

**Comment by stubailo** [comment]
> Unrelated, but one disadvantage of the convention to put fragments as static members of the component is you can't reference these from static propTypes {} member functions.

Could you file this on `react-apollo` please?

The above looks like a bug, where we are accidentally appending the fragment twice.

Can you reproduce some of your code here? Where you are running the query and setting the variables, etc?

**Comment by richburdon** [comment]
Thanks @stubailo:

Child Item

```
export const ItemFragments = {
  item: new Fragment(gql`
    fragment ItemFragment on Item {
      id
      title
    }
  `)
};

export class Item extends React.Component {
  static propTypes = {
    item: ItemFragments.item.propType
  };
  render() {
    let { item } = this.props;
    return (
      <div>{ this.props.title }</div>
    );
  }
}
```

Parent List:
```
const GetItemsQuery = gql`
  query GetItems($text: String, $offset: Int, $count: Int) { 
    items(text: $text, offset: $offset, count: $count) {
      id
      ...ItemFragment
    }
  }
`;

  graphql(GetItemsQuery, {
    options: ({ text }) => {
      return {
        fragments: ItemFragments.item.fragments(), // WAS EXPECTING TO PASS IN text HERE.
        variables: {
          text: text,
          offset: 0,
          count: 10
        }
      };
    },

    props({ ownProps: { text }, data: { loading, items, fetchMore } }) {
      return {
        loading,
        text,
        items,
        fetchMoreItems() {
          return fetchMore({
            variables: {
              text: text,
              offset: items.length
            },
            updateQuery: (previousResult, { fetchMoreResult }) => {
              return _.assign({}, previousResult, {
                items: [...previousResult.items, ...fetchMoreResult.data.items]
              });
            }
          });
        }
      }
    }
  }),

```

**Comment by richburdon** [comment]
Temporary workaround:

```
networkInterface.use([{
  applyMiddleware({ request }, next) {
    let definitions = {};
    request.query.definitions = _.filter(request.query.definitions, (definition) => {
      let name = definition.name.value;
      if (definitions[name]) {
        return false;
      } else {
        definitions[name] = true;
        return true;
      }
    });

    next();
  }
}]);
```

**Comment by stubailo** [comment]
Great workaround.

If you could submit a failing test that would be super helpful as well: https://github.com/apollostack/apollo-client/blob/dc8cc0063d93a8c15a90771bc88183bc058b5e1b/test/fetchMore.ts

**Comment by richburdon** [comment]
OK. I'll try to wrap my head around that and have a stab at it (might be closer to Monday though).

Thanks for the comments.

**Comment by stubailo** [comment]
Yeah, we won't be able to get to a fix before Monday either. Thanks for the really detailed report!

**Comment by yurigenin** [comment]
I had the same issue. The workaround for me was to explicitly specify the same query and fragments in the call to fetchMore.

**Comment by stubailo** [comment]
Looks like the fix would be in here: https://github.com/apollostack/apollo-client/blob/372bca9881b374caf56d0cbd0795a88dcc13ec0f/src/queries/getFromAST.ts#L150

**Comment by helfer** [comment]
@stubailo are you suggesting that we "fix" this by checking whether a fragment with the same name already exists on the query? The other fix would be not to attach it twice in the first place. I'm leaning towards a simple fix for now, because we'll soon want to refactor how we transform queries anyway.

**Comment by jnwng** [comment]
@helfer see #906 for a (fairly) simple fix for this particular issue (which is not attaching it twice in the first place)

warnings about fragment name collisions are already being surfaced in  [createFragment](https://github.com/apollostack/apollo-client/blob/master/src/fragments.ts#L39)

**Comment by helfer** [comment]
Solved in #913
